sympy/functions/combinatorial/factorials.py

```diff
@@ -1029,12 +1064,50 @@ def _eval_expand_func(self, **hints):
         else:
             return binomial(*self.args)
 
+    def _eval_rewrite_as_Piecewise(self, n, k, **kwargs):
+        from sympy.concrete.products import Product
+        from sympy import gamma, Abs, nan
+        knk = lambda x: Or(Eq(k, x), Eq(n - k, x))
+        isint = lambda x: Eq(x, floor(x))
+        nint = lambda x: Ne(x, floor(x))
+        pos = lambda x: And(Ne(x, 0), Eq(x/Abs(x), 1))
+        neg = lambda x: And(Ne(x, 0), Eq(x/Abs(x), -1))
+        inf = lambda x: And(Ne(x, 0), Eq(1/x, 0))
+        for i in igen():
+            if not n.has(i) and not k.has(i):
+                break
+        return Piecewise(
+            (1, Or(Eq(n, k), knk(0))),
+            (n, knk(1)),
+            (zoo, And(neg(n), isint(n), nint(k))),
+            (nan, Or(inf(n), inf(k))),
+            (0, Or(
+                knk(-k),
```

Comment:
```suggestion
```

Reply:
I removed it locally. I'll push it soon.
but after this change,
```python3
>>> b.subs(x, 0)
Piecewise((1, Eq(y, 0)), (0, True))
```
☝️  changed to this 👇 
```python3
>>> b.subs(x, 0)
Piecewise((1, Eq(y, 0)), (0, (Eq(y, -1) | Eq(y, 1) | Ne(y, 0)) & (Eq(y, -1) | Eq(y, 1) | Eq(y, floor(y))) & (Eq(y, -1) | Eq(y,
 1) | Eq(y/Abs(y), -1))), (Product((i - y)/i, (i, 1, y)), Eq(y, floor(y))), (1/(gamma(1 - y)*gamma(y + 1)), True))
```

Reply:
That simple expression is the result of the bad condition `knk(-k)` which is true since, with `n=0`, is `Or(k==-k, 0-k==-k) -> Or(k==-k,True) -> True`. Since that True is itself in an Or predicate for the 0 result the predicate is True so the Piecewise collapses to 2 args: one for 1 and one for 0. 

Hmm...if n is 0 then the answer *is* zero if k > 0; are there other conditions? Maybe the knk(-k) should be removed from where it is and a new condition be added here:
```python
And(
                    And(isint(n), isint(k)),
                    Or(Eq(n, 0), <---------------------------
                         And(pos(n*k), pos(k - n)),
                         And(pos(n), neg(k)))))
```
I can't check this now. Basically we have to get the case of n=0 -> 0 into the Piecewise if it's not already there.